DESIGN.md

```diff
@@ -0,0 +1,178 @@
+Most Unix environments will use UTF-8
+and most Unix-y utilities will produce textual logs with reasonably sized lines.
+
+`GetLogs` prioritizes latency over throughput.
+It will try to send multiple log lines in a single message, but this is not guaranteed.
+In general, new log messages are sent as soon as possible, without additional buffering.
+
+### `GetStatus`
+
+`GetStatus` returns the status of the process with the given ID.
+Currently, the response only indicates whether the process has exited,
+in which case the response will contain either the signal number
+(if the process has been killed by a signal), or the exit code otherwise.
+
+### `Stop`
+
+`Stop` stops the process with the given ID if it is still running.
+It first tries to let the program terminate gracefully by sending a SIGTERM.
+If the process is still running after a set amount of time, it will send a SIGKILL.
+This mirrors the behavior of many process managers, for example,
+[systemd](https://stackoverflow.com/questions/42978358/how-systemd-stop-command-actually-works).
+
+It is a no-op to issue a `Stop` command for a process that has already been finished.
+However, trying to `Stop` a process that is in the process of being stopped returns an error.
+
+### Error Conditions
+
+To indicate errors, built-in gRPC status codes have been chosen for simplicity's sake.
+A human-readable error message will be provided in the response.
+
+Examples of error conditions:
+
+* Invalid arguments to `Exec` (empty array of strings)
+* Error while spawning the process in `Exec` (e.g. the specified binary does not exist)
+* Invalid process ID given to `GetLogs`, `GetStatus`, `Stop`
+* Issuing `Stop` to a process when another `Stop` operation is in progress for that process.
+
+## Security
+
+### Transport
+
+`paasd` uses TLS for transport security.
+The clients only trust servers that produce a valid certificate
+signed by *server certificate authority (CA)*.
+
+As there is no requirement for compatibility with any legacy systems,
+it is planned to support only the strongest cipher suites,
+i.e. category "A" as outlined by [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/TLS_Cipher_String_Cheat_Sheet.html#table-of-the-ciphers-and-their-priority-from-high-1-to-low-eg-18).
+All of these ciphers provide both perfect forward secrecy
+and authenticated encryption.
+We choose to support multiple ciphers,
+as this provides independent upgrade-paths for both client and server,
+in case any of these ciphers get deprecated.
+
+### Authentication and authorization
+
+`paasd` uses mTLS to authenticate the clients via client certificates.
+Upon connection, it is validated that the client has a valid certificate,
+issued by the *client CA*.
+
+We assume that every client has a unique certificate
+and so use the certificates to uniquely identify clients.
+Internally, a cryptographic hash of the client certificate could be used
+to have a short, strong identifier for a client.
```

Comment:
Certificates allow you encode data in them, could you encode data within the certificate itself that would help you identify the user? Seems simpler than keeping track of hashes.

Reply:
That sounds like a good idea. Would using the subject's common name to identify be reasonable?

Reply:
That works for me, please update the design document and merge.